src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,191 @@
+            for (int i = 0; i < rows.Length; ++i)
+            {
+                Row row = rows[i];
+                Assert.Equal(1, row.Size());
+                Assert.Equal(expected[i], row.GetAs<string>(0));
+            }
+        }
+
+        /// <summary>
+        /// UDF that returns Array type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnAsArrayType()
+        {
+            // UDF with return as ArrayType throws a following exception:
+            // Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
+            // ---> System.NotImplementedException: The method or operation is not implemented.
+            // at Microsoft.Spark.Sql.Row.Convert() in Microsoft.Spark\Sql\Row.cs:line 169
+            // at Microsoft.Spark.Sql.Row..ctor(Object[] values, StructType schema) in Microsoft.Spark\Sql\Row.cs:line 34
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in Microsoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<string, string[]>(
+                str => new string[] { str, str + str });
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());
+
+            //Show() works here. See the example below.
+            _df.Select(udf(_df["name"])).Show();
+        }
+
+        /// <summary>
+        /// UDF that takes in Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithMapType()
+        {
+            // UDF with MapType throws a following exception:
+            // [] [] [Error] [TaskRunner] [0] ProcessStream() failed with exception: System.InvalidCastException: Unable to cast object of type 'System.Collections.Hashtable' to type 'System.Collections.Generic.IDictionary`2[System.String,System.String]'.
+            // at Microsoft.Spark.Sql.PicklingUdfWrapper`2.Execute(Int32 splitIndex, Object[] input, Int32[] argOffsets) in Microsoft.Spark\Sql\PicklingUdfWrapper.cs:line 44
+            // at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.SingleCommandRunner.Run(Int32 splitId, Object input) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 239
+            // at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 139
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    dict => dict.Count.ToString());
+
+            DataFrame df = _df.WithColumn("tempName", Map(_df["name"], _df["name"]));

```

Comment:
tempName?  How about something like NameIdsMap and create a map from Name => Ids

Reply:
Changed!